### PR TITLE
record+routing/route: add AMP record

### PR DESCRIPTION
**Pull request by cfromknecht**
This PR adds the currently proposed AMP record from https://github.com/lightningnetwork/lightning-rfc/pull/658, and
implements the sanity checks when encoding a particular `route.Hop`. For now, this could will
remain dead until a number of other components are in place, but this serves as a starting point to
begin implementing the remainder of the proposal that depends on these primitive definitions
before ultimately enabling it externally.

Related to #3939.

### CODE REVIEW

File: record/amp.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+// PayloadSize returns the size this record takes up in encoded form.
+func (a *AMP) PayloadSize() uint64 {
+	return 32 + 32 + tlv.SizeTUint16(a.childIndex)
+}
+
+// String returns a human-readble description of the amp payload fields.
+func (a *AMP) String() string {
+	return fmt.Sprintf("root_share=%x set_id=%x child_index=%d",
```
Comment (Roasbeef) [inline record/amp.go]: 👍

File: routing/route/route.go [hunk 2]
```diff
@@ -168,6 +176,18 @@ func (h *Hop) PackHopPayload(w io.Writer, nextChanID uint64) error {
 		}
 	}
 
+	// If an AMP record is destined for this hop, ensure that we only ever
```
Comment (Roasbeef) [inline routing/route/route.go]: 👍

**Comment by cfromknecht** [comment]
fixed the linter errors, build now passes

**Comment by joostjager** [comment]
Does merging this now really make things easier? It doesn't look like this code has conflict potential. In general I think it is better to merge living code, unless we have for example two prs that build onto this. But not sure if that is already the case.

**Pull request merged**